### PR TITLE
[SPARK-50290][SQL] Add a flag to disable DataFrameQueryContext creation

### PR DESCRIPTION
**Pull request by vladimirg-db**
### What changes were proposed in this pull request?

Add a new `spark.sql.dataFrameQueryContext.enabled` flag to disable the `DataFrameQueryContext` creation.

### Why are the changes needed?

`DataFrameQueryContext` creation requires a stack trace. Stack trace collection has a non-trivial perf overhead, some users might want to disable this feature.

`spark.sql.dataFrameQueryContext.enabled` == `true`:

image

`spark.sql.dataFrameQueryContext.enabled` == `false`:

image

### Does this PR introduce _any_ user-facing change?

No, the default is still to collect the query context.

### How was this patch tested?

Added a new test case.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionAnsiErrorsSuite.scala
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionAnsiErrorsSuite.scala:414]: hmm, why do we still provide this empty context?

Comment (vladimirg-db) [reply sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionAnsiErrorsSuite.scala:414]: The context is always passed to `divideByZeroError`:
image

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionAnsiErrorsSuite.scala
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionAnsiErrorsSuite.scala:407]: Could you move this test to the more specific suite: `QueryContextSuite`, please.

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionAnsiErrorsSuite.scala
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionAnsiErrorsSuite.scala:414]: I see, thanks.

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionAnsiErrorsSuite.scala
Comment (vladimirg-db) [reply sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionAnsiErrorsSuite.scala]: Done.

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

**Review by zsxwing — APPROVED** [review]
LGTM

**Review by HyukjinKwon — APPROVED** [review]
cc @itholic

**Review by itholic — APPROVED** [review]
Looks good, and I think we should add a corresponding flag for PySpark as well.
Let me made a followup after merging this PR.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by itholic** [comment]
fyi: I create follow-up ticket and working on it: SPARK-50310

**Pull request closed without merging**
